packages/engine/crawlers/solana/transactionHistory.ts

```diff
@@ -0,0 +1,530 @@
+	ParsedTransactionWithMeta,
+	PartiallyDecodedInstruction,
+} from '@solana/web3.js';
+import { PublicKey } from '@solana/web3.js';
+import type { Collectible, Token, TokenAccount } from '@walless/core';
+import { Networks } from '@walless/core';
+
+import { historyActions } from './../../state/history/index';
+import { getMetadata, solMint } from './metadata';
+import type { SolanaContext } from './shared';
+
+const gasilonConfigs = await fetch(`${GASILON_ENDPOINT}`, {
+	method: 'GET',
+	headers: {
+		'Content-Type': 'application/json',
+	},
+});
+
+const FEE_PAYER = (await gasilonConfigs.json()).feePayer;
+export interface Transaction {
+	id: string;
+	signature: string;
+	network: Networks;
+	type: 'sent' | 'received';
+	status: 'success' | 'pending' | 'failed';
+	sender: string;
+	receiver: string;
+	token: Token | Omit<Collectible, 'account' | 'collectionId'>;
+	fee: number;
+	tokenForFee?: Token;
+	preBalance?: number;
+	postBalance?: number;
+	amount: number;
+	date: Date;
+}
+
+const getConfirmation = async (
+	solanaContext: SolanaContext,
+	signature: string,
+) => {
+	const result = await solanaContext.connection.getSignatureStatus(signature, {
+		searchTransactionHistory: true,
+	});
+	return result.value?.confirmationStatus;
+};
+
+const isAbstractionFee = (meta: ParsedTransactionWithMeta['meta']) => {
+	const tokenBalances = meta?.postTokenBalances;
+
+	for (const tokenBalance of tokenBalances ?? []) {
+		if (tokenBalance.owner === FEE_PAYER) {
+			return true;
+		}
+	}
+
+	return false;
+};
+
+const getTransactionTokenForFee = async (
+	instruction: ParsedInstruction | PartiallyDecodedInstruction,
+	meta: ParsedTransactionWithMeta['meta'],
+	solanaContext: SolanaContext,
+) => {
+	if (isAbstractionFee(meta)) {
+		const tokenBalances = meta?.postTokenBalances;
+
+		for (const tokenBalance of tokenBalances ?? []) {
+			if (!tokenBalance.owner) continue;
+
+			if (tokenBalance.owner === FEE_PAYER) {
+				const network = Networks.solana;
+				const account: TokenAccount = {
+					mint: tokenBalance.mint,
+					owner: tokenBalance.owner,
+					address: (instruction as ParsedInstruction).parsed?.info.source,
+					balance: tokenBalance.uiTokenAmount.uiAmount?.toString() || '',
+					decimals: tokenBalance.uiTokenAmount.decimals,
+				};
+				const metadata = await getMetadata(solanaContext, tokenBalance.mint);
+
+				const token: Token = {
+					network,
+					account,
+					metadata,
+				};
+
+				return token;
+			}
+		}
+	}
+
+	return {
+		network: Networks.solana,
+		account: {
+			mint: solMint,
+			owner: 'system',
+			address: solMint,
+			balance: '',
+			decimals: 9,
+		},
+		metadata: await getMetadata(solanaContext, solMint),
+	};
+};
+
+const getTransactionFee = (
+	transaction: ParsedTransactionWithMeta,
+	tokenForFee: Token,
+) => {
+	let fee = (transaction.meta?.fee ?? 5000) / 10 ** 9;
+
+	if (isAbstractionFee(transaction.meta)) {
+		const instruction = transaction.transaction.message.instructions[0];
```

Comment:
It's not sure if getting an abstract transaction by the first instruction is correct. Gasilon supports that this instruction can be in any position, in case this transaction comes from an external source, it might cause problems

Reply:
can you suggest me another way to deal with this?